### PR TITLE
cql3: Reorganize expr::to_restriction

### LINKED ISSUE #10631: Invalid IN restriction is reported as a '=' restriction
In the cql-pytest `cassandra_tests/validation/operations/select_single_column_relation_test.py::testInvalidCollectionNonEQRelation`
we have the query
```
SELECT * FROM %s WHERE c = 0 AND b IN (?)
```
Where b is a `set<int>`, Scylla reports the error:
>  Collection column 'b' (set<int>) cannot be restricted by a '=' relation

Obviously, this is a 'IN' relation, not a '=' relation. Cassandra prints 'IN' here and so did Scylla until very recently (when I translated this test, it used to pass).

### PR DESCRIPTION
**Pull request by cvybhu**
This PR introduces improvements to `expr::to_restriction` and prepares the validation part for restriction classes removal.

`expr::to_restriction` is currently used to take a restriction from the WHERE clause, prepare it, perform some validation checks and finally convert it to an instance of the restriction class.

Soon we will get rid of the restriction class.

In preparation for that `expr::to_restriction` is split into two independent parts:
* The part that prepares and validates a binary_operator
* The part that converts a binary_operator to restriction

Thanks to this split getting rid of restriction class will be painless, we will just stop using the second part.

`to_restriction.cc` is replaced by `restrictions.hh/cc`. In the future we can put all the restriction expressions code there to avoid clutter in `expression.hh/cc`.

This change made it much easier to fix #10631, so I did that as well.

Fixes: #10631

### CODE REVIEW

**Comment by cvybhu** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1193/

I think the CI failure wasn't because of faulty code. The error message says it failed to clone seastar because it ran out of disk space:
```
23:29:48  hudson.plugins.git.GitException: Command "git submodule update --init --recursive seastar" returned status code 1:
23:29:48  stdout: Submodule path 'seastar': checked out '7d8d846b26743601e145adc32db87d7f86a25d3b'
23:29:48  
23:29:48  stderr: Cloning into '/jenkins/workspace/releng/Scylla-CI@5/scylla/seastar'...
23:29:48  Submodule 'dpdk' (git@github.com:scylladb/dpdk) registered for path 'seastar/dpdk'
23:29:48  Cloning into '/jenkins/workspace/releng/Scylla-CI@5/scylla/seastar/dpdk'...
23:29:48  error: unable to create file drivers/mempool/dpaa/dpaa_mempool.h: No space left on device
23:29:48  error: unable to create file drivers/mempool/dpaa/meson.build: No space left on device
23:29:48  error: unable to create file drivers/mempool/dpaa/rte_mempool_dpaa_version.map: No space left on device
23:29:48  fatal: cannot create directory at 'drivers/mempool/dpaa2': No space left on device
23:29:48  fatal: Unable to checkout '0f486994114dbc68da07e7345674f7d83b54e0fc' in submodule path 'seastar/dpdk'
23:29:48  fatal: Failed to recurse into submodule path 'seastar'
```
Is there some way for me to retrigger CI? I could rebase but that's a bit heavy handed.

**Comment by nyh** [comment]
Yes, you can go to the CI link, and choose "rebuild". I'll do it now.

File: cql3/expr/to_restriction.cc
Comment (nyh) [inline cql3/expr/to_restriction.cc]: If we had a bug here, it probably means our tests do not cover this case :-( i.e., that "a IS NOT 3" should be refused when creating a materialized view - not cause a weird failure later when adding rows.
Would be good to add a regression test when we find such bugs.

File: cql3/expr/restrictions.cc
Comment (nyh) [inline cql3/expr/restrictions.cc:250]: I'm not familiar with this code, so I have to ask: Why do we bother with this optimization (and risk of introducing bugs in it - do we have tests for single-item IN?)? I understand that a single-element IN is perhaps slightly less efficient than =, but is the difference substantial?

**Review by nyh — APPROVED** [review]
Looks good, but I left a couple of questions for your consideration - and suggested to add another small regression test.

**Comment by cvybhu** [comment]
> CI state FAILURE - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1193/

Same thing, no space left on device :C

**Comment by avikivity** [comment]
@benipeled please fix ASAP

**Comment by benipeled** [comment]
> @benipeled please fix ASAP

fixed yesterday (at least temporary :/
early today I retrigger the build

File: cql3/expr/to_restriction.cc
Comment (cvybhu) [reply cql3/expr/to_restriction.cc]: Added a test, although it turns out that `IS NOT NULL` is checked earlier at the grammar level.

File: cql3/expr/to_restriction.cc
Comment (avikivity) [inline cql3/expr/to_restriction.cc]: I started work to convert IS NOT NULL to be a unary_expression (so we can remove expr::null), but hit snags. Maybe this work can help it along.

**Comment by avikivity** [comment]
Looks good.

Please move the Fixes tag to the commit that actually fixes it, since the main goal of the series isn't to fix the bug.

**Comment by nyh** [comment]
> Looks good.
> 
> Please move the Fixes tag to the commit that actually fixes it, since the main goal of the series isn't to fix the bug.

But if I understand correctly, the fix can only be done after the reorganizational patches before - it's not like you can backport just the one specific patch that fixes the bug. No?

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
> > > Looks good.
> > > Please move the Fixes tag to the commit that actually fixes it, since the main goal of the series isn't to fix the bug.
> > 
> > 
> > But if I understand correctly, the fix can only be done after the reorganizational patches before - it's not like you can backport just the one specific patch that fixes the bug. No?
> 
> Right, but it's odd if "reorganize to_restriction" fixes a bug.

It wouldn't be the first time that a bug fix comes only after a major code reorganization, which makes it hard to impossible to backport. But I guess you're right that the bug fix can indded be considered as a separate patch, which comes **after** the code reorganization - it could even even been a separate PR. If somebody wants to backport only that patch, they will fail and need to figure out which additional patches they need to backport.

Anyway, I don't feel strongly either way, so I'll let you and @cvybhu decide where to put the "FIXES" :-)

**Comment by cvybhu** [comment]
Moved `Fixes` to the commit where it fits the best.
Had to swap order of two commits to make it work.

Before that there was one more change that needed to be done in order for the test to pass and I got confused where to put the Fixes.

**Comment by nyh** [comment]
Thanks, I'll wait for the CI and commit this (hoping it won't crash as usual on no space on device...).

**Comment by cvybhu** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1238/

```
20:30:04  E   cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="unconfigured table varcharmaptable"
```
This might be some kind of race condition where the CI tries to perform queries without waiting for schema agreement, I will run it again.

**Comment by cvybhu** [comment]
> I will run it again.

I'm not sure if I did the right thing.

First I clicked on the failed build and chose to rebuild it. I didn't enter any arguments, only the PR number was filled (automatically). But then I noticed that there are the wrong commits in build description, so I assumed that I made a mistake and that I should've filled in the other fields, so I aborted it.
image

Then I tried to fill in PR, user and branch to point the CI to the right commits (without the separator field, what's that?). But it didn't like that and failed with:
```
ERROR: Redundant parameters have been passed, please use pull-request OR fork parameters but not both
```

So now I'm rebuilding again with just the PR argument. I hope it's checking the right thing. It's this build: https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1243/

**Comment by psarna** [comment]
@cvybhu try rebasing onto current master, it will retrigger the CI build and maybe also get rid of the issue

**Comment by benipeled** [comment]
> First I clicked on the failed build and chose to rebuild it. I didn't enter any arguments, only the PR number was filled (automatically). But then I noticed that there are the wrong commits in build description, so I assumed that I made a mistake and that I should've filled in the other fields, so I aborted it.

@cvybhu You did it right - hit the rebuild button - and you can see your commit and PR on the build page
image

> Then I tried to fill in PR, user and branch to point the CI to the right commits (without the separator field, what's that?). But it didn't like that and failed with:

You should point the job to PR **or** your fork-branch (useful when you want you to verify your patch before the PR submission)

You shouldn't guess these options - I hope to make it simpler soon on https://github.com/scylladb/scylla/issues/10980

**Comment by avikivity** [comment]
> > CI state **`ABORTED`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1243/
> 
> huh I don't know why it got aborted, I will just rebase onto master.

Maybe related to Jenkins migration this weekend.

**Comment by avikivity** [comment]
The signed-off-by tags are broken in the commits.

**Comment by avikivity** [comment]
> The signed-off-by tags are broken in the commits.

The From: tags too.

**Comment by cvybhu** [comment]
> The signed-off-by tags are broken in the commits.

Changed `Signed-off-by` and `Author` to have `Jan Ciolek <jan.ciolek@scylladb.com>` instead of `cvybhu <jan.ciolek@scylladb.com>`

> The From: tags too.

Do you mean `Author:` or is there some other hidden tag?

**Comment by cvybhu** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1280/

```
15:08:04  In file included from <built-in>:1:
15:08:04  In file included from ./db/per_partition_rate_limit_extension.hh:11:
15:08:04  In file included from ./db/per_partition_rate_limit_options.hh:15:
15:08:04  ./schema.hh:625:13: error: no type named 'per_partition_rate_limit_options' in namespace 'db'
15:08:04          db::per_partition_rate_limit_options _per_partition_rate_limit_options;
15:08:04          ~~~~^
15:08:04  ./schema.hh:818:15: error: no type named 'per_partition_rate_limit_options' in namespace 'db'
15:08:04      const db::per_partition_rate_limit_options& per_partition_rate_limit_options() const {
15:08:04            ~~~~^
15:08:04  ./schema.hh:887:17: error: cannot deduce return type 'const auto &' for function with no return statements
15:08:04      const auto& dropped_columns() const {
15:08:04                  ^
15:08:04  ./schema.hh:891:17: error: cannot deduce return type 'const auto &' for function with no return statements
15:08:04      const auto& collections() const {
```

Looks like something unrelated to this PR

**Comment by avikivity** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1280/
> 
> ```
> 15:08:04  In file included from <built-in>:1:
> 15:08:04  In file included from ./db/per_partition_rate_limit_extension.hh:11:
> 15:08:04  In file included from ./db/per_partition_rate_limit_options.hh:15:
> 15:08:04  ./schema.hh:625:13: error: no type named 'per_partition_rate_limit_options' in namespace 'db'
> 15:08:04          db::per_partition_rate_limit_options _per_partition_rate_limit_options;
> 15:08:04          ~~~~^
> 15:08:04  ./schema.hh:818:15: error: no type named 'per_partition_rate_limit_options' in namespace 'db'
> 15:08:04      const db::per_partition_rate_limit_options& per_partition_rate_limit_options() const {
> 15:08:04            ~~~~^
> 15:08:04  ./schema.hh:887:17: error: cannot deduce return type 'const auto &' for function with no return statements
> 15:08:04      const auto& dropped_columns() const {
> 15:08:04                  ^
> 15:08:04  ./schema.hh:891:17: error: cannot deduce return type 'const auto &' for function with no return statements
> 15:08:04      const auto& collections() const {
> ```
> 
> Looks like something unrelated to this PR

Looks like a bad rebase. You have to fix it if you want it merged.

**Comment by cvybhu** [comment]
> > > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1280/
> > 
> > 
> > ```
> > 15:08:04  In file included from <built-in>:1:
> > 15:08:04  In file included from ./db/per_partition_rate_limit_extension.hh:11:
> > 15:08:04  In file included from ./db/per_partition_rate_limit_options.hh:15:
> > 15:08:04  ./schema.hh:625:13: error: no type named 'per_partition_rate_limit_options' in namespace 'db'
> > 15:08:04          db::per_partition_rate_limit_options _per_partition_rate_limit_options;
> > 15:08:04          ~~~~^
> > 15:08:04  ./schema.hh:818:15: error: no type named 'per_partition_rate_limit_options' in namespace 'db'
> > 15:08:04      const db::per_partition_rate_limit_options& per_partition_rate_limit_options() const {
> > 15:08:04            ~~~~^
> > 15:08:04  ./schema.hh:887:17: error: cannot deduce return type 'const auto &' for function with no return statements
> > 15:08:04      const auto& dropped_columns() const {
> > 15:08:04                  ^
> > 15:08:04  ./schema.hh:891:17: error: cannot deduce return type 'const auto &' for function with no return statements
> > 15:08:04      const auto& collections() const {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looks like something unrelated to this PR
> 
> Looks like a bad rebase. You have to fix it if you want it merged.

The compilation errors are in files completely unrelated to my PR, I couldn't have broken them on rebase 0_o.
I will rebase onto the newest master, but I'm not convinced if that's the issue.

**Comment by avikivity** [comment]
Well, master compiles. So it's your series that doesn't compile.

**Comment by avikivity** [comment]
Maybe it's a dev-headers test? Try `ninja dev-headers`.

**Comment by benipeled** [comment]
> Maybe it's a dev-headers test? Try `ninja dev-headers`.

Yes, you can see on the console output that the failure is following `ninja -j 32 dev-headers`

**Comment by psarna** [comment]
@cvybhu: https://github.com/scylladb/scylla/pull/10986#issuecomment-1180331039

**Pull request closed without merging**
